templates/analysis.yml

```diff
@@ -0,0 +1,137 @@
+# External requirements:
+# ----------------------
+# In order to guard secrets, two variable groups must be present in the
+# ADO library and called in the pipeline YAML file before using this
+# template.
```

Comment:
from this template's POV, why do they need to be in these two groups? isn't that a detail for the pipeline to figure out? the real requirement seems to be that the six variables exist.
you could make those 6 parameters too, then the template doesn't even care about variables at all. that's perhaps more up for debate.

Reply:
My initial design had the secrets being passed as parameters, but the secrets are not available at build-time when the parameters are assigned static values, so it would pass in string literals that looked like '$(sonarOrg)' instead of expanding the variables.
You are correct that they could be in one group, but I liked the organization of having them separate since they have separate purposes.  They don't need to be two variable groups, so I could avoid stating that explicitly, however, I wanted it to be clear to the developers what is going on in this specific use case, so I decided to err on being overly specific so that the usage in this project would be clear.
At one point I had each microservice's key and name spelled-out, but then I couldn't pass them dynamically to the template, so I had to find a way to use static values and set them with dynamic values so that the secrets remained hidden, but the key elements that were different were passed in as static parameters.

Reply:
To clarify about the two groups, I anticipate adding the discord variables to pipelines that do not need the sonarCloud values.

Reply:
>it would pass in string literals that looked like '$(sonarOrg)' instead of expanding the variables.

i would expect this to work fine. the job and steps the agent executes would have that text, and at that point they would be expanded into the variable's value at that point in time, if in fact there was a variable with that name. (iirc, text like that remains unchanged if the variable doesn't exist)

you can document everything that someone needs to know about at the surface level on the pipeline file. the templates are meant to be more reusable. two groups is fine in the pipeline file.

Reply:
Revisiting this after learning more about variables from other tasks to get the template up and running, I found that the issue was in my formatting of the variables and not in passing them to the template.
I have refactored the template to use external secrets passed as parameters.
My previous conclusions were erroneous.  I misdiagnosed the issue due to apparently misunderstanding some answers I found online.